gcovr/gcov.py

```diff
@@ -424,6 +452,7 @@ def parse_tag_line(
 
             branch_cov = self.coverage.line(self.lineno).branch(branch_index)
             branch_cov.count += count
+            branch_cov.executed = executed
```

Comment:
Hi @latk 
Thank you for your review! I need more time to think about other comments and do some fixes :), but I have my doubts for this points.

> It seems the executed field would be redundant with count? If so, we don't need it. The BranchCoverage class might still have a read-only property for convenience, if that helps with reports.

When I have `.gcov` output like this:
```sh
call    0 returned 100%
branch  1 taken 100% (fallthrough)
branch  2 taken 0% (throw)
        1:   25:  total = 0;
```

How `count` variable from `gcov.py` will be set for this line: `branch  2 taken 0% (throw)`?
```Python
branch_cov.count += count
```
I think it will be equal to 0. Branch was executed and taken with 0%.
For this example:
```sh
call    0 returned 100%
branch  1 taken 100% (fallthrough)
branch  2 never executed
```    
the `count` variable is also equal to 0.
We are not able to disctinguist between `branch  2 taken 0% (throw)` and `branch  2 never executed`?

> As implemented, the two fields can actually go out of sync. Assume a branch in a header file that is included in two compilation units. In unit A, the branch is taken two times. We have count = 2, executed = True. In unit B, the branch is never executed. This code will now result in count = 2, executed = False which is a contradiction.

Can you show me an example with this branch in the header file?

Reply:
I've looked at the gcov source code to understand where the “never executed” concept comes from:

https://github.com/gcc-mirror/gcc/blob/08a4adcf2b6ded2fea97195c715757df61a23395/gcc/gcov.c#L2861-L2867

```c
      if (arc->src->count)
	fnotice (gcov_file, "branch %2d taken %s%s", ix,
		 format_gcov (arc->count, arc->src->count, -flag_counts),
		 arc->fall_through ? " (fallthrough)"
		 : arc->is_throw ? " (throw)" : "");
      else
	fnotice (gcov_file, "branch %2d never executed", ix);
```

In the gcov data model, there are blocks and control flow arcs between blocks. A block corresponds to a chunk of assembly code terminated by a jump to a different block, e.g. with a conditional or call.

Here we have an `arc` with a count, and the `arc->src` block from which this arc originates. If `arc->src->count` is zero, that means the control flow instruction like `jle` was never executed. Incidentally, this also sidesteps division by zero errors when calculating the percentage. The information `not executed` implies `not covered` and is a stronger statement. But is it really useful to know whether the branching instruction / conditional was executed? In most cases, that information would already be included in line coverage.

---

As a practical example, I've created some files with a conflict in the attached [never-executed.zip](https://github.com/gcovr/gcovr/files/5815938/never-executed.zip). It consists of a main() function that links two compilation units test-a.c and test-b.c. They both include the test.h header which contains a static function that will be compiled separately for each compilation unit.

File `test-a.c##test.h.gcov` shows coverage resulting from test-a.c calling this function:

```
        2:    1:inline static int brancher(int x) {
        2:    2:    if (x > 0) {
branch  0 taken 2 (fallthrough)
branch  1 taken 0
        2:    3:        return x * 2;
        -:    4:    } else {
    #####:    5:        return 0;
        -:    6:    }
        -:    7:}
```

In contrast, `test-b.c##test.h.gcov` shows that the function was compiled but not called:

```
    #####:    1:inline static int brancher(int x) {
    #####:    2:    if (x > 0) {
branch  0 never executed
branch  1 never executed
    #####:    3:        return x * 2;
        -:    4:    } else {
    #####:    5:        return 0;
        -:    6:    }
        -:    7:}
```

I believe that whether your code considers the branches on line 2 as executed or not executed would depend on the order in which these two gcov files are processed. If A is processed before B, the branches would be marked as never executed.

The correct approach would be to default the `executed` field to False, and only update it to True. It can never go from True to False. I.e. the value should be combined with `or`. But, as discussed above, I'm still not convinced that the info about whether a branching operator was executed is meaningful information for a coverage report. I've seen that the upstream gcov JSON format (which we don't parse yet) does not show this info on a branch level, but merely notes whether a line contains any unexecuted block.

Reply:
Thinking about this some more: I don't want to make changes that *prevent* future evolution of gcovr. Committing to showing certain information in reports is a strong constraint. If (at some point in the future) we were to ingest upstream gcov JSON as a coverage data source, we would not be able to show the same information about executed calls & branches. Being forced to stick with the textual gcov format will be substantial technical debt. The current textual format is bad enough because it fails to show “throw” status for never executed branches.

If this branch-level `executed` information has a strong use case we can still include it, but make it's use opt-in with a command line flag that can be documented as being compatible with only certain coverage sources.